File: img_preprocess/img_preprocess.py
import os
import cv2
import numpy as np

IMAGE_DIR = "image"
OUTPUT_DIR = "processed_images"

def img_preprocess(img_path):
    img = cv2.imread(img_path, 0) # 以灰度模式读取图像
    if img is None:
        return None
    sampled = img[::4, ::4] # 每隔4个像素采样一次，降低分辨率
    return (np.floor(sampled / 16) * 16).astype(np.uint8) # 将像素值降低到16的倍数


def main():
    os.makedirs(OUTPUT_DIR, exist_ok=True)

    for img_file in os.listdir(IMAGE_DIR):
        img_path = os.path.join(IMAGE_DIR, img_file)
        processed_img = img_preprocess(img_path)
        if processed_img is None:
            continue
        processed_img_path = os.path.join(OUTPUT_DIR, img_file)
        cv2.imwrite(processed_img_path, processed_img)


if __name__ == "__main__":
    main()

File: img_preprocess/img_preprocess.m
ROOT_DIR = fileparts(fileparts(mfilename('fullpath')));
IMAGE_DIR = fullfile(ROOT_DIR, "image");
OUTPUT_DIR = fullfile(ROOT_DIR, "processed_images");

main(IMAGE_DIR, OUTPUT_DIR);

function out = preprocess_image(img_path)
    img = imread(img_path);
    img = rgb2gray(img);
    sampled = img(1:4:end, 1:4:end);
    out = uint8(floor(double(sampled) / 16) * 16);
end

function main(image_dir, output_dir)
    if ~exist(output_dir, 'dir')
        mkdir(output_dir);
    end

    files = dir(fullfile(image_dir, '*.jpg'));
    for k = 1:numel(files)
        img_path = fullfile(image_dir, files(k).name);
        processed_img = preprocess_image(img_path);
        out_path = fullfile(output_dir, files(k).name);
        imwrite(processed_img, out_path);
    end
end